src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,485 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief Reverse-mode automatic differentiation.
+ *
+ * Now only supports differentiating one function in the IRModule with one dataflow block
+ * with respect to the only return value of the function, which needs to be scalar.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+#include "../op/make_op.h"
+
+namespace tvm {
+namespace relax {
+
+// a tool class for GradientMutator
+// visit the forward bindings and generate the backward bindings
+class BackwardBindingGenerator : public ExprVisitor {
+ public:
+  explicit BackwardBindingGenerator(const BlockBuilder& builder) : builder_(builder) {}
+
+  /*!
+   * \brief Generate the backward bindings for the corresponding GradientMutator
+   *
+   * \param forward_block The forward dataflow block
+   * \param require_grads The arguments to dif
+   * \param post The expression with rewritten inputs.
+   */
+  Expr Generate(const DataflowBlock& forward_block, const Array<Var>& require_grads,
```

Comment:
Shall we make Generate a static function, and construct BackwardBindingGenerator in it?